### DISCUSSION TITLE
🚀使用mosdns替代openclash默认的dns分流

### DISCUSSION THREAD

**Discussion by nagaame** [nagaame/giscus-discussions#7] [Announcements]
# https://kuokuo.io/2022/03/24/openclash-with-mosdns/

在我的openwrt下openclash它自己的DNS分流有时候可能会出现DNS污染的问题,然后我们其实可以使用其他的DNS转发工具来替代openclash的DNS转发功能, 比如smartdns和mosdns, 个人认为smartdns 所以选择了比较简单和上手容易的mosdns

https://kuokuo.io/2022/03/24/openclash-with-mosdns/

**Comment by wangjianpro999** [comment]
但是依然有个问题：openclash中默认监听的dns端口不用改？
DNS监听端口7874

Dnsmasq的常规设置里的DNS转发端口会随着MosDNS的启用而改变。
所以您的教程里应该怎么处理这些问题？

**Comment by nagaame** [comment]
> @wangjianpro999
> 但是依然有个问题：openclash中默认监听的dns端口不用改？
> DNS监听端口7874
> 
> Dnsmasq的常规设置里的DNS转发端口会随着MosDNS的启用而改变。
> 所以您的教程里应该怎么处理这些问题？

不用, 修改好openclash的DNS设置之后, 启动openclash, 它会自己修改dnsmasq, 让openclash作为dnsmasq的上游的, 不知道这么说能不能解决你的疑惑

**Comment by wangjianpro999** [comment]
我用的固件是404大佬的，mosdns监听5335，openclash则监听7874，启动任何一个都会改变Dnsmasq的设置，所以这两个只能选其一，而不改变openclash的监听端口，则这两者启动先后顺序决定了Dnsmasq启用的端口号。这个您没有解答我的疑惑。

壳壳中的宇宙 ***@***.***> 于 2022年3月27日周日 14:12写道：

**Comment by nagaame** [comment]
> @wangjianpro999
> 我用的固件是404大佬的，mosdns监听5335，openclash则监听7874，启动任何一个都会改变Dnsmasq的设置，所以这两个只能选其一，而不改变openclash的监听端口，则这两者启动先后顺序决定了Dnsmasq启用的端口号。这个您没有解答我的疑惑。
> 
> 壳壳中的宇宙 ***@***.***> 于 2022年3月27日周日 14:12写道：
> 
> > @wangjianpro999 <https://github.com/wangjianpro999>
> > 但是依然有个问题：openclash中默认监听的dns端口不用改？
> > DNS监听端口7874
> >
> > Dnsmasq的常规设置里的DNS转发端口会随着MosDNS的启用而改变。
> > 所以您的教程里应该怎么处理这些问题？
> >
> > 不用, 修改好openclash的DNS设置之后, 启动openclash, 它会自己修改dnsmasq,
> > 让openclash作为dnsmasq的上游的, 不知道这么说能不能解决你的疑惑
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <nagaame/giscus#4>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AS7OYSGXIJB6BXMPVU3YBJDVB733NANCNFSM5RQ6HOPA>
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
> 

当openclash没有启动的时候, `dnsmasq`, 转发设置默认是空的或者转发到其他, 你可以在这里填一个国内的DNS, 以防止`openclash`启动的时候没有获取到服务器的地址, `openclash`的DNS设置, 填`mosdns`的本地地址`udp://127.0.0.1:5335`(包括nameserver和fallback) , 然后我们分别启动mosdns和openclash, openclash启动成功之后, 它会自动修改dnsmasq的转发为`udp://127.0.0.1:7874`到openclash这里, dnsmasq转发这里不用修改成转发到mosdns. 它们都正常工作后, dnsmasq的上游是openclash, openclash的dns上游又是mosdns.

**Reply by duzink** [reply]
特意来感谢,果然评论区出精华,DNS转发7874这个逻辑问题困扰了我很久,这条问答豁然开朗.

**Comment by UNDESAIN** [comment]
感觉你的配置写的更简单，借用了。现在用passwall+mosdns（你提供的配置），刚用一会感觉挺稳的。之前的配置，有的时候貌似比如打开咸鱼也会提示开启了VPN，有时正常，有些报错。

**Reply by UNDESAIN** [reply]
特意留言表示感谢！

**Reply by nagaame** [reply]
好的, 有问题随时反馈ヽ(￣▽￣)و

**Comment by kyle-wu-1996** [comment]
我使用的是MosDNS +openclash，使用流程按照您的教程设置，但是在MosDNS 中使用内置预设可以正常启动，换成自定义配置后 就在启动和停止间反复跳动，请问这怎么弄，是配置文件的问题吗？

**Reply by nagaame** [reply]
检查一下有没有`/etc/init.d/mosdns`文件, 没有的话, 去[仓库](https://github.com/QiuSimons/openwrt-mos/tree/master/luci-app-mosdns/root/etc/init.d)拷贝一个回来, 放到原来的位置, 然后使用命令`chmod +x mosdns`添加执行权限, 然后重新`/etc/init.d/mosdns enable` 和 `/etc/init.d/mosdns reload`一下试试, 还不确定是mosdns核心新版本还是openwrt-mos导致的问题

**Comment by dqbpqdpb** [comment]
建议把 - 'ext:./geosite.dat:apple-cn'加入query_is_local_domain，这样苹果的服务可以走国内服务器。
国内的DNS我觉得用阿里和腾讯2个就好。
阿里的DOH https://dns.alidns.com/dns-query和腾讯的DOH https://1.12.12.12/dns-query
国外就谷歌加CF2个就好，没啥毛病。更简单。
感谢博主的模板！

**Reply by nagaame** [reply]
感谢建议🥰

**Comment by teel-writ-ros** [comment]
设置好，已成功。感谢！

另外能否考虑再加个adguard home + openclash + mosdns的配置教程？谢谢！

**Reply by nagaame** [reply]
待定，还不确定加不加，我研究看看😂

**Reply by teel-writ-ros** [reply]
自己研究了一下写了一个教程，引用了一下你的帖子。https://www.right.com.cn/forum/thread-8227823-1-1.html

**Reply by nagaame** [reply]
> 自己研究了一下写了一个教程，引用了一下你的帖子。[right.com.cn/forum/thread-8227823-1-1.html](https://www.right.com.cn/forum/thread-8227823-1-1.html)

可以的，完全没问题，因为我现在暂时还不打算配置adguard home，mosdns和openclash的一些配置可能还有优化的空间，感谢

**Reply by duncan198475** [reply]
caesarjing在恩山看了你的帖子 这周末还没折腾好 我已经魔怔了 是否能加上我qq65202001远程辅导一下 谢谢

**Comment by dqbpqdpb** [comment]
根据 https://github.com/IrineSistiana/mosdns/issues/206 的建议，geoip文件可以换成https://github.com/Loyalsoldier/geoip  这个项目里面的cn.dat，只有150K，性能效率大大提升。
反正geoip的唯一用处一般就是判断一下是不是CN的IP。

**Reply by nagaame** [reply]
感谢建议，我替换一下

**Comment by timsims** [comment]
配置里防广告那段被注释了，是没效果吗还是什么原因？

**Reply by nagaame** [reply]
去广告有效果的，不过效果不明显吧，你可以使用mosdns去广告，也可以使用adguard home去广告，其实我推荐后者

**Comment by vps2me-google** [comment]
你好，博主，现在mosdns已升到3.9.0.1了，你的自定义配置无法运行，只用默认了

**Reply by nagaame** [reply]
看看log有什么报错吗, 我看3.9版本是有一些不兼容的更新, 你把 hosts_map相关的东西都删除, 重新运行试试

**Comment by duncan198475** [comment]
你好，博主，我应用自定义模板显示未运行。然后你说测试mosdns端口是否通的那个代码 在哪里输入啊 我是新手 谢谢

**Reply by nagaame** [reply]
你可以使用ssh工具登录到软路由之后, 使用命令行操作, 也可以进到软路由后台, 菜单找到TTYD终端执行命令

**Reply by duncan198475** [reply]
用putty登录 输入$ netstat -a | grep 5335 显示找不到文件。。。。

**Reply by nagaame** [reply]
不要输入$符号哦, 直接输入`netstat -a | grep 5335`就可以了, $符号在linux系统表示默认用户, 如果是root用户是#符号

**Reply by duncan198475** [reply]
博主可以加我qq65202001协助我设置吗 我已经快被折腾疯了 谢谢

**Reply by duncan198475** [reply]
我直接winscp进入op的etc/mosdns文件夹，把cus_config.yml打开，把你的代码复制粘贴进去，然后重启op，直接网断了。。。

**Reply by duncan198475** [reply]
我用putty登录 然后输入netstat -a | grep 5335 没有任何反应。。。

**Comment by duncan198475** [comment]
并且我mosdns插件的logo是空的 是没有运行吗

**Reply by nagaame** [reply]
有可能没有成功运行吧, 可以使用命令检查端口或者使用`ps`命令检查运行的进程

**Comment by gdanmac** [comment]
info	coremain/run.go:102	working directory changed	{"path": "/etc/mosdns/"}
fatal	coremain/run.go:136	failed to parse config file	{"error": "1 error(s) decoding:\n\n* '' has invalid keys: plugin"}

更新了新版 配置文件提示报错了

**Reply by nagaame** [reply]
可以去在[这里](https://verytoolz.com/yaml-formatter.html)检查一下yaml文件的格式, 是不是正常

**Reply by gdanmac** [reply]
Valid YAML
提示正常。。https://op.supes.top/packages/x86_64/ 更新这个里面版本 58版本

**Reply by gdanmac** [reply]
mosdns_385aec0-57_x86_64.ipk 这个版本 用配置文件没问题。

**Reply by gdanmac** [reply]
mosdns ver: 385aec0 这个版本

**Reply by nagaame** [reply]
可能是新版本一些不兼容的问题吧, 错误应该是解析yaml出问题了

**Reply by nagaame** [reply]
也可能是mosdns新版本不在需要plugin关键字, 我还得查一下试试, 如果旧版本没问题的话, 先用着旧版本也可以的

**Comment by Pacio** [comment]
更新了最新版本的mosdns，这次无法启动了，报错提示如下:

2022-06-19T08:02:57.661Z        info    v3/main.go:157  current working directory: /etc/mosdns/
2022-06-19T08:02:57.662Z        info    v3/main.go:223  mosdns ver: v3.9.0
2022-06-19T08:02:57.662Z        info    v3/main.go:224  arch: arm64, os: linux, go: go1.18.3
2022-06-19T08:02:57.662Z        info    coremain/run.go:65      loading config  {"file": "def_config.yaml"}
2022-06-19T08:02:57.665Z        fatal   coremain/run.go:37      failed to load config   {"error": "failed to parse the config f                                                                                                                                          ile: yaml: line 165: mapping values are not allowed in this context"}

**Reply by nagaame** [reply]
尝试换成自定义配置, 其实就是cus_config.yaml文件, 然后使用我的配置文件试试

**Reply by Pacio** [reply]
> 尝试换成自定义配置, 其实就是cus_config.yaml文件, 然后使用我的配置文件试试

现在无论默认还是自定义都启动不了~好像作者最近在搞alpha版本测试，所以不稳定，我现在用的smartdns，过段时间等稳定了再用吧

**Reply by nagaame** [reply]
了解了, 你可以尝试使用下边的内容替换`/etc/init.d/mosdns`这个文件的内容, 替换前可以先备份源文件
```bash
#!/bin/sh /etc/rc.common
# Copyright (C) 2020 IrineSistiana

START=99
USE_PROCD=1

#####  ONLY CHANGE THIS BLOCK  ######
PROG=/usr/bin/mosdns # where is mosdns
RES_DIR=/etc/mosdns/ # resource dir / working dir / the dir where you store ip/domain lists
CONF=./cus_config.yaml   # where is the config file, it can be a relative path to $RES_DIR
#####  ONLY CHANGE THIS BLOCK  ######

start_service() {
  procd_open_instance
  procd_set_param command $PROG -dir $RES_DIR -c $CONF

  procd_set_param user root
  procd_set_param stdout 1
  procd_set_param stderr 1
  procd_set_param respawn "${respawn_threshold:-3600}" "${respawn_timeout:-5}" "${respawn_retry:-5}"
  procd_close_instance
  echo "mosdns is started!"
}

reload_service() {
  stop
  sleep 2s
  echo "mosdns is restarted!"
  start
}

```

**Comment by ccwssy** [comment]
博主,有4.0版本的配置么

**Reply by nagaame** [reply]
最近比较忙, 我看看周末有没有时间更新一下😂

**Reply by nagaame** [reply]
V4版本的配置已经更新了，需要清空`/etc/mosdns`文件夹, `/etc/init.d/mosdns`文件等，然后重新安装`luci-app-mosdns`和v4版本的mosdns core，使用v4版本的配置文件应该可以用了

**Comment by gdanmac** [comment]
查到原因了 是自动更新geoip 的问题。 自己手动更新就没事。。

**Comment by duzink** [comment]
mosdns当前配置能正常运行"ssh下端口5335正常工作"那么分流就一定是正常的

ccwssy ***@***.***> 于2022年7月20日周三 23:20写道：

**Comment by liangxyz** [comment]
楼主你好，请问我openclash设置fakeip模式就不能上网，设置redir-host模式就能上网怎么回事呀？

**Reply by nagaame** [reply]
理论上fakeip速度更快, 但是redir-host兼容性更好, 如果你没有特殊要求的话, 可以暂时继续使用redir-host模式, 另外, fakeip模式会影响到ddns, 注意要做ip白名单排除

**Comment by wangjianpro999** [comment]
楼主答复不了，要群主答复😆

liangxyz ***@***.***> 于 2022年8月7日周日 19:20写道：

**Comment by yyysuo** [comment]
有个疑问，看您的配置，首先是Openclash根据自己的分流规则先对dns进行了分流，分成国内、国外2组dns请求，又统一汇总到了mosdns，其实等于把openclash自己的dns分流功能给废弃了。然后mosdns又把汇总的dns按自己的分流规则再次分流，解析后传递给openclash，假如某个域名，mosdns分流到了国外，但是在openclash分流里面是国内的，就会出现对国外cdn友好的IP从国内直连，影响速度；反之，如果某个域名，mosdns分流到了国内，在openclash分流里面是国外的，那么解析的ip是对国内cdn友好，但是会走代理绕一圈访问，也会影响速度。

按最优的结果，如果openclash的dns分流与mosdns的分流结果完全一致， 那为什么不直接用openclash自己的分流呢，直接把mosdns的上游dns填到openclash里面就行了啊，那mosdns的意义只剩下乐观缓存和屏蔽规则了，这些功能在科学软件、Adguard Home、smartdns中都有的。

**Reply by nagaame** [reply]
你说的有道理, 当初搞这个方案的时候, 是因为单纯使用openclash分流, 有时候会出现DNS污染的问题, 我不确定是openclash的问题还是我设置openclash不当的问题, 然后我使用mosdns作为前置的dns解析之后, 发现问题消失了, 而且可以自己做一些自定义的dns设置, 所以就这样配合使用下去了, 如果你觉得太麻烦, 单独使用openclash也是可以的

**Reply by yyysuo** [reply]
有好处就不怕复杂，我折腾得更复杂，之所以研究mosdns就是想把它也加到套娃里面，哈哈。

https://www.right.com.cn/forum/thread-8247205-1-1.html

**Reply by nagaame** [reply]
> 有好处就不怕复杂，我折腾得更复杂，之所以研究mosdns就是想把它也加到套娃里面，哈哈。
> 
> [right.com.cn/forum/thread-8247205-1-1.html](https://www.right.com.cn/forum/thread-8247205-1-1.html)

很赞哈哈, mosdns, 资源占用不高, 功能也挺多的, 应该还有其他玩法

**Reply by gdanmac** [reply]
以前用过openclash的dns分流的老是出问题，不知道现在修复了没有，只从用了mosdns 就没安逸了。。。

**Reply by nagaame** [reply]
> 以前用过openclash的dns分流的老是出问题，不知道现在修复了没有，只从用了mosdns 就没安逸了。。。

mosdns v4版本不知道稳定了没, 我一直在用3.9版本, 挺稳定的

**Reply by yyysuo** [reply]
> > 以前用过openclash的dns分流的老是出问题，不知道现在修复了没有，只从用了mosdns 就没安逸了。。。
> 
> mosdns v4版本不知道稳定了没, 我一直在用3.9版本, 挺稳定的

我也把mosdns加到套娃里面了，用的就是v4，用起来感觉还不错，ssrp没有geosite的分流，用mosdns就算是加上了。
![全板dns流程+ MOSDNS](https://user-images.githubusercontent.com/10081071/196206966-bd0190f1-0a84-410a-800d-42e11aa9f31e.jpg)

**Reply by nagaame** [reply]
> > > 以前用过openclash的dns分流的老是出问题，不知道现在修复了没有，只从用了mosdns 就没安逸了。。。
> > 
> > 
> > mosdns v4版本不知道稳定了没, 我一直在用3.9版本, 挺稳定的
> 
> 我也把mosdns加到套娃里面了，用的就是v4，用起来感觉还不错，ssrp没有geosite的分流，用mosdns就算是加上了。 ![全板dns流程+ MOSDNS](https://user-images.githubusercontent.com/10081071/196206966-bd0190f1-0a84-410a-800d-42e11aa9f31e.jpg)

可以, 很强👍

**Reply by Journalist-HK** [reply]
我的用法是解决传统黑名单模式下不代理的域名的DNS污染问题，配合 passwall 使用非常稳定。我的配置：[mosdns-config](https://github.com/Journalist-HK/mosdns-config)

**Reply by nagaame** [reply]
> 我的用法是解决传统黑名单模式下不代理的域名的DNS污染问题，配合 passwall 使用非常稳定。我的配置：[mosdns-config](https://github.com/Journalist-HK/mosdns-config)

passwall满足你的需要的话, 也是可以的, 一开始我也是用passwallo(*￣︶￣*)o

**Comment by wswcx** [comment]
大佬，请教一下1.4.2版本的luci-app-mosdns在哪下载，在您的页面里貌似只有luci-app-mosdns的1.2链接~

**Reply by nagaame** [reply]
这里有下载, 你可以看看 https://github.com/sbwml/luci-app-mosdns/releases

**Reply by wswcx** [reply]
收到，非常感谢~~

**Comment by gdanmac** [comment]
V4的规则开启IPV6 后报错， 还有就是出新版的能更新一下吗？

servers:
  - exec: main_sequence
    listeners:
      - protocol: udp
        addr: "[::1]:5335"
      - protocol: tcp
        addr: "[::1]:5335"
      - protocol: udp
        addr: "0.0.0.0:5335"
      - protocol: tcp
        addr: "0.0.0.0:5335"

**Reply by gdanmac** [reply]
addr: "0.0.0.0:5335"改了127.0.0.0了 。。有啥区别

**Reply by nagaame** [reply]
0.0.0.0是指任意地址，监听这个地址的话，任意主机都可以连接到本机，127.0.0.1是本地环回地址，监听这个地址只能在本机访问

**Reply by gdanmac** [reply]
我用"0.0.0.0:5335"的时候 是打不开IPV6（addr: "[::1]:5335"）的这 ，说端口冲突。。

**Reply by Journalist-HK** [reply]
你写法有误，v4v6只要一条":5335"就行了

**Reply by gdanmac** [reply]
你好 ，我使用是这个配置文件由@gdyan2001提供，这个文件是打开了IPV6。还是说没打开IPV6，使用IPV4也能获取解析V6的？

**Reply by nagaame** [reply]
现阶段的配置, 解析和使用ipv6可能有问题, 还需要看你的openwrt的ipv6相关组件和ipv6的设置是否正常, 暂时不建议增加解析ipv6的配置

**Comment by Journalist-HK** [comment]
clash内置dns的fallback功能就是国内外ip分流，mosdns完全可以替代，clash配置里不需要有fallback dns。我目前遇到的问题是fake-ip和redir-host哪一个更适合配合mosdns工作

**Reply by nagaame** [reply]
redir-host兼容性更好, fake-ip速度更快, 不管你使用这两个哪个模式, 都建议打开tun模式

**Comment by JoeM78** [comment]
请问大佬会更新v5版本的配置吗

**Reply by nagaame** [reply]
或许吧^_^我周末才有空，到时候升级到V5的配置

**Reply by JoeM78** [reply]
root@iStoreOS:~# netstat -a | grep 5335
tcp        0      0 :::5335                 :::*                    LISTEN      
udp        0      0 0.0.0.0:53359           0.0.0.0:*         

大佬我按教程    返回的结果是这样  只有这两条  然后udp还是53359

**Reply by nagaame** [reply]
端口监听这里改一下, 就使用这样就行了

```yaml

servers:
  - exec: main_sequence
    listeners:
      # - protocol: udp
      #   addr: "[::1]:5335"
      # - protocol: tcp
      #   addr: "[::1]:5335"
      - protocol: udp
        addr: "0.0.0.0:5335"
      - protocol: tcp
        addr: "0.0.0.0:5335"
```

**Reply by JoeM78** [reply]
复制进去显示没有待修改…之前好像也是这么写的

**Reply by JoeM78** [reply]
是不是因为我的界面多了一个dns转发  我给它打勾了
截图：
https://s2.loli.net/2023/02/24/mQGx3EO94hqotaC.png

**Reply by nagaame** [reply]
dns转发不用打开，除非你知道这个是什么用途，然后如果修改不成功可以直接修改源文件，位置在这里，自定义的配置文件文件名是`config_custom.yaml`
```bash
root@OpenWrt:/etc/mosdns# pwd
/etc/mosdns
root@OpenWrt:/etc/mosdns# ls
cn.dat                   config.yaml              config_custom.yaml       config_custom.yaml-opkg  geoip.dat                geosite.dat              rule

```

修改完之后重载一下mosdns服务

**Comment by orthwang** [comment]
openclashDNS设置 增加两个DNS服务器127.0.0.1：5335 是否只勾选这两个选项 取消其他dns的勾选

**Reply by nagaame** [reply]
是的，openclash直接使用mosdns的监听端口就可以了

**Reply by orthwang** [reply]
谢谢 教程及解答

**Comment by nemome66** [comment]
v5的配置可以更新下吗？

**Reply by nagaame** [reply]
周末我更新一下

**Reply by dayuanpan** [reply]
同球V5配置

**Reply by nemome66** [reply]
大佬，来催更了。

**Reply by gdanmac** [reply]
同球V5配置，要和现在V4一样的功能。 真好用啊

**Reply by ccwssy** [reply]
给个github上的两个V5参考配置,自己可以参考配置改一下就可以用了。
https://github.com/QiuSimons/openwrt-mos/blob/master/dat/def_config_v5.yaml
https://github.com/sbwml/luci-app-mosdns/blob/v5/luci-app-mosdns/root/etc/mosdns/config_custom.yaml

我所使用的自动更新脚本，我是用的docker版：
#!/bin/sh
wget --no-check-certificate https://raw.githubusercontent.com/Loyalsoldier/geoip/release/text/cn.txt -O  
 你的目录/config/mosdns/data/cn_ip.txt
wget --no-check-certificate https://raw.githubusercontent.com/Loyalsoldier/v2ray-rules-dat/release/reject-list.txt -O  你的目录/config/mosdns/data/reject_list.txt
wget --no-check-certificate https://github.com/MartialBE/v2dat_rule/releases/latest/download/geosite_cn.txt -O  你的目录/config/mosdns/data/geosite_cn.txt
wget --no-check-certificate https://github.com/MartialBE/v2dat_rule/releases/latest/download/geolocation-no-cn.txt -O  你的目录/config/mosdns/data/geolocation-no-cn.txt
docker restart mosdns

说明：
cn.txt  国内IP
reject-list.txt 广告或者黑名单
geosite_cn.txt 国内域名
geolocation-no-cn.txt  非国内域名

**Comment by fengchev** [comment]
大佬，我用了您的v4版本的配置，很好用，但是我这里有个问题，有时mosdns会突然失效，没法上网，我是在旁路由部署的，但是mosdns出问题的时候，切到主路由也会没法上网，看了下主路由后台，显示openwrt的连接数直接超过上w，关了mosdns就没事了，请问这是我的openwrt版本还是mosdns有问题吗

**Reply by nagaame** [reply]
还不是很确定，也可能是openwrt的其他组件影响的，毕竟涉及到dns的请求和处理，你可以看看mosdns的log和openwrt自己的系统log和内核log

**Reply by gdanmac** [reply]
找的原因了，是被移动宽带 劫持了 抢答，我用ros 指定线路。。

**Reply by gdanmac** [reply]
回错了， 感觉你这个问题 要去查日志，还有叫什么内存泄露。导致。。

**Comment by gdanmac** [comment]
最近开了V6后老是遇到这些错误，不知怎么解决？

2023-07-17T04:03:31.542Z	warn	dnspod	upstream err	{"query": "invalid\\ IP. IN A 42348 161 ::ffff:127.0.0.1", "addr": "https://doh.pub/dns-query"}
2023-07-17T04:03:31.546Z	warn	dnspod	upstream err	{"query": "invalid\\ IP. IN AAAA 25748 162 ::ffff:127.0.0.1", "addr": "https://doh.pub/dns-query"}

2023-07-17T04:05:21.008Z	warn	cloudflare	upstream err	{"query": "explodie.org. IN AAAA 46655 279 ::ffff:127.0.0.1", "addr": "https://cloudflare-dns.com/dns-query"}
2023-07-17T04:05:21.008Z	warn	cloudflare	upstream err	{"query": "explodie.org. IN AAAA 46655 280 ::ffff:127.0.0.1", "addr": "https://cloudflare-dns.com/dns-query"}
2023-07-17T04:06:02.721Z	warn	forward_remote.node_0	sequence failed	{"query": "iaat-tech.com.cn. IN A 23866 324 ::ffff:127.0.0.1", "sequence": 2, "error": "opening quic connection to quic://dns-unfiltered.adguard:853: CRYPTO_ERROR (0x12a): x509: certificate is valid for dns.adguard.com, dns-family.adguard.com, dns-unfiltered.adguard.com, not dns-unfiltered.adguard"}
2023-07-17T04:06:12.598Z	warn	forward_remote.node_0	sequence failed	{"query": "tracker.kuroy.me. IN A 45044 327 ::ffff:127.0.0.1", "sequence": 2, "error": "opening quic connection to quic://dns-unfiltered.adguard:853: CRYPTO_ERROR (0x12a): x509: certificate is valid for dns.adguard.com, dns-family.adguard.com, dns-unfiltered.adguard.com, not dns-unfiltered.adguard"}

**Reply by yyysuo** [reply]
直接改成ip不行吗，为什么一定要用域名呢。

**Reply by nagaame** [reply]
看log像是你的doh 的解析被污染了，所以没有解析到这个doh的服务器，你可以尝试换成dns server换成ip或者让代理服务器通过dns流量

**Reply by gdanmac** [reply]
看到 域名有些少了com          - addr: "quic://dns-unfiltered.adguard"” "https://dns.google/dns-query"

**Reply by nagaame** [reply]
这个com不用管它，域名本来就是这样的，主要是协议不同，这些相当于是加密的dns解析协议

**Reply by gdanmac** [reply]
找的原因了，是被移动宽带 劫持了 抢答，我用ros 指定线路。。

**Reply by nagaame** [reply]
那就是这个原因了，移动就是墙中墙，一般电信不会污染这些dns服务器域名，但是移动只要是敏感的域名和ip都封完了

**Reply by gdanmac** [reply]
https://doh.pub/dns-query
我改成https://120.53.53.53/dns-query
形式了 然后添加了ipv6的解析。。

**Reply by nagaame** [reply]
HTTPS一般来说是不能用ip形式的,你这样的话,可能会提示证书错误,导致无法连接到doh服务器

**Reply by gdanmac** [reply]
走电信线路 还是有一堆错误。。

**Reply by nagaame** [reply]
本地解析和远端解析你应该没有搞混吧？本地解析的话，用国内的dns，然后就可以获取远端的dns服务器的地址

**Reply by ccwssy** [reply]
dnspod的doh服务器是可以直接用ip的形式来连接的，SSL证书是包含了ip的地址的

IP Address=120.53.53.53
IP Address=119.28.28.99
IP Address=1.12.12.12
DNS Name=*.dot.pub
DNS Name=dot.pub

**Reply by nagaame** [reply]
cloudflare dns 似乎也支持，doh的ip形式地址是这个`https://1.1.1.1/dns-query`

**Reply by nagaame** [reply]
@gdanmac 你可以尝试换cf的ip形式doh的dns服务器地址`https://1.1.1.1/dns-query`

**Reply by ccwssy** [reply]
cloudflare dns 需要走梯子才能有效果，直连一样会有很多报错

**Reply by gdanmac** [reply]
cloudflare 域名有走梯子。我在adg 上Bootstrap DNS 服务器改成了运营商的地址有好转一些。但是会出现eam err	{"query": "musicstatus-apac-s.edge-itunes-apple.com.akadns.net. IN HTTPS 41212 90688 ::ffff:127.0.0.1", "addr": "119.29.29.29"}
2023-07-27T08:54:04.964Z	warn	dnspod	upstream err	{"query": "musicstatus-apac-s.edge-itunes-apple.com.akadns.net. IN HTTPS 41212 90688 ::ffff:127.0.0.1", "addr": "tls://dot.pub"}
2023-07-27T08:54:04.965Z	warn	dnspod	upstream err	{"query": "musicstatus-apac-s.edge-itunes-apple.com.akadns.net. IN HTTPS 41212 90688 ::ffff:127.0.0.1", "addr": "https://120.53.53.53/dns-query" 这些错

**Reply by nagaame** [reply]
这些是警告，不算错误，如果不影响使用，dns没有问题的话，可以不用管

**Comment by k3soft912** [comment]
感谢大佬的分享，刚接触这个还不知道怎么搞，顺便问下没有V5版本的自定义配置吗？

**Reply by nagaame** [reply]
可以看看最新评论区

**Reply by k3soft912** [reply]
评论区看过了！没看到大佬的V5配置！所以问一下

**Reply by nagaame** [reply]
@Journalist-HK  和 @ccwssy 他们的配置就是V5的。。。

**Reply by k3soft912** [reply]
感谢大佬提示！

**Reply by yyysuo** [reply]
直接用sbwml大佬的luci就好了，手撸没必要。  https://github.com/sbwml/luci-app-mosdns/releases

**Comment by Journalist-HK** [comment]
分享一下自用v5配置：[mosdns-config](https://github.com/Journalist-HK/mosdns-config) 目前公网内网都稳定使用。

**Comment by RookieTerry** [comment]
你好，请问这样设置以后，openclash的dns劫持还需要打开吗？

**Reply by nagaame** [reply]
不需要哦, 转发到mosdns了
